### PR TITLE
add started timestamp to workflow instance table

### PR DESCRIPTION
**Pull request by efonsell**

### CODE REVIEW

File: nflow-engine/src/main/resources/scripts/db/update-5.5.0-x/h2.update.ddl.sql [hunk 1]
```diff
@@ -0,0 +1 @@
+alter table nflow_workflow add started timestamp;
```
Comment (jsyrjala) [inline nflow-engine/src/main/resources/scripts/db/update-5.5.0-x/h2.update.ddl.sql]: Is it possible to have UPDATE commands that populate `started` field based on data in nflow_action table?

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,4 +1,8 @@
-## 5.5.1-SNAPSHOT (future release)
+## 6.0.0-SNAPSHOT (future release)
+
+** Highlights **
+- BREAKING CHANGE: add started timestamp to workflow instance table
```
Comment (jsyrjala) [inline CHANGELOG.md]: Add something about why these changes were added. e.g Faster queries when there is a lot of actions/state variable changes

File: nflow-engine/src/main/resources/scripts/db/update-5.5.0-x/h2.update.ddl.sql (hunk 1 above)
Comment (efonsell) [reply nflow-engine/src/main/resources/scripts/db/update-5.5.0-x/h2.update.ddl.sql]: Added

File: CHANGELOG.md (hunk 2 above)
Comment (efonsell) [reply CHANGELOG.md]: Updated

File: nflow-engine/src/main/resources/scripts/db/update-5.5.0-x/db2.update.ddl.sql [hunk 3]
```diff
@@ -1 +1,4 @@
 alter table nflow_workflow add started timestamp(3);
+
+update nflow_workflow w, (select workflow_id, min(execution_start) as started from nflow_workflow_action group by workflow_id) a
+  set w.started = a.started where w.id = a.workflow_id
```
Comment (jsyrjala) [inline nflow-engine/src/main/resources/scripts/db/update-5.5.0-x/db2.update.ddl.sql]: Semicolon missing from the end of the command.

File: nflow-engine/src/main/resources/scripts/db/mysql.legacy.create.ddl.sql [hunk 4]
```diff
@@ -108,6 +109,7 @@ create table if not exists nflow_archive_workflow (
   retries int not null default 0,
   created timestamp not null,
   modified timestamp not null,
+  started timestamp,
```
Comment (jsyrjala) [inline nflow-engine/src/main/resources/scripts/db/mysql.legacy.create.ddl.sql]: nflow_archive_workflow doesn't have `null`, nflow_workflow has `null`. Shouldn't the both fields be identical?

File: nflow-engine/src/main/resources/scripts/db/update-5.5.0-x/db2.update.ddl.sql [hunk 5]
```diff
@@ -0,0 +1,4 @@
+alter table nflow_workflow add started timestamp(3);
```
Comment (jsyrjala) [inline nflow-engine/src/main/resources/scripts/db/update-5.5.0-x/db2.update.ddl.sql]: xxx.create.ddl.sql files also add `started` field to nflow_archive_workflow table but these xxx.update.ddl.sql files do not add `started` field

Comment (jsyrjala) [reply nflow-engine/src/main/resources/scripts/db/update-5.5.0-x/db2.update.ddl.sql]: Also should we have UPDATE to fill started fields also for nflow_archive_workflow here?

File: nflow-engine/src/main/resources/scripts/db/update-5.5.0-x/db2.update.ddl.sql (hunk 3 above)
Comment (efonsell) [reply nflow-engine/src/main/resources/scripts/db/update-5.5.0-x/db2.update.ddl.sql]: Added

File: nflow-engine/src/main/resources/scripts/db/mysql.legacy.create.ddl.sql (hunk 4 above)
Comment (efonsell) [reply nflow-engine/src/main/resources/scripts/db/mysql.legacy.create.ddl.sql]: Yes, fixed

File: nflow-engine/src/main/resources/scripts/db/update-5.5.0-x/db2.update.ddl.sql (hunk 5 above)
Comment (efonsell) [reply nflow-engine/src/main/resources/scripts/db/update-5.5.0-x/db2.update.ddl.sql]: Yes, fixed

File: CHANGELOG.md [hunk 6]
```diff
@@ -2,6 +2,7 @@
 
 **Highlights**
 - Support non-enum WorkflowStates to enable, for example, Kotlin workflow definitions by extending AbstractWorkflowDefinition.
+- Added `started` timestamp to workflow instance table (requires database update)
```
Comment (jsyrjala) [inline CHANGELOG.md]: Needs a rebase after 5.6.0 release and moving these changes to future release

File: nflow-engine/src/main/resources/scripts/db/update-5.5.0-x/db2.update.ddl.sql [hunk 7]
```diff
@@ -0,0 +1,7 @@
...
+alter table nflow_archive_workflow add started timestamp(3);
```
Comment (jsyrjala) [inline nflow-engine/src/main/resources/scripts/db/update-5.5.0-x/db2.update.ddl.sql]: Directory names for update scripts need to be updated after 5.6.0 release.

File: nflow-rest-api-jax-rs/src/test/java/io/nflow/rest/v1/jaxrs/WorkflowInstanceResourceTest.java [hunk 8]
```diff
@@ -3,6 +3,7 @@
 import static com.nitorcreations.Matchers.hasField;
 import static io.nflow.engine.workflow.instance.WorkflowInstanceAction.WorkflowActionType.externalChange;
 import static java.util.Arrays.asList;
+import static java.util.Collections.emptySet;
```
Comment (jsyrjala) [inline nflow-rest-api-jax-rs/src/test/java/io/nflow/rest/v1/jaxrs/WorkflowInstanceResourceTest.java]: WorkflowInstanceResourceTest fails on travis-ci.

https://travis-ci.org/NitorCreations/nflow/jobs/533785975

This could be helpful in fixing tests.
`lenient().when(...).thenReturn(..)`

File: nflow-engine/src/main/java/io/nflow/engine/internal/dao/WorkflowInstanceDao.java [hunk 9]
```diff
@@ -323,16 +322,16 @@ public void updateWorkflowInstanceAfterExecution(WorkflowInstance instance, Work
         updateWorkflowInstanceWithTransaction(instance, action, childWorkflows, workflows, changedStateVariables);
       }
     } else {
-      updateWorkflowInstance(instance);
+      updateWorkflowInstance(instance, action.executionStart);
     }
   }
 
-  public int updateWorkflowInstance(WorkflowInstance instance) {
+  public int updateWorkflowInstance(WorkflowInstance instance, DateTime started) {
```
Comment (gmokki) [inline nflow-engine/src/main/java/io/nflow/engine/internal/dao/WorkflowInstanceDao.java]: Do not pass the `started` argument in since the `instance.started` already should contain the correct value. And if it doesn't then the caller should update it to be correct before updating the instance. There is nothing special in `started` compared to the other timestamp fields.

File: nflow-engine/src/main/java/io/nflow/engine/internal/executor/WorkflowStateProcessor.java [hunk 10]
```diff
@@ -178,7 +177,7 @@ private void rescheduleUnknownWorkflowType(WorkflowInstance instance) {
     logger.warn("Workflow type {} not configured to this nFlow instance - rescheduling workflow instance", instance.type);
     instance = new WorkflowInstance.Builder(instance).setNextActivation(now().plusMinutes(unknownWorkflowTypeRetryDelay))
         .setStatus(inProgress).setStateText("Unsupported workflow type").build();
-    workflowInstanceDao.updateWorkflowInstance(instance);
+    workflowInstanceDao.updateWorkflowInstance(instance, null);
```
Comment (gmokki) [inline nflow-engine/src/main/java/io/nflow/engine/internal/executor/WorkflowStateProcessor.java]: The `null` will clear the current start time from the db.

File: CHANGELOG.md (hunk 6 above)
Comment (efonsell) [reply CHANGELOG.md]: Done

File: nflow-engine/src/main/resources/scripts/db/update-5.5.0-x/db2.update.ddl.sql (hunk 7 above)
Comment (efonsell) [reply nflow-engine/src/main/resources/scripts/db/update-5.5.0-x/db2.update.ddl.sql]: Done

File: nflow-rest-api-jax-rs/src/test/java/io/nflow/rest/v1/jaxrs/WorkflowInstanceResourceTest.java (hunk 8 above)
Comment (efonsell) [reply nflow-rest-api-jax-rs/src/test/java/io/nflow/rest/v1/jaxrs/WorkflowInstanceResourceTest.java]: Fixed

File: nflow-engine/src/main/java/io/nflow/engine/internal/dao/WorkflowInstanceDao.java (hunk 9 above)
Comment (efonsell) [reply nflow-engine/src/main/java/io/nflow/engine/internal/dao/WorkflowInstanceDao.java]: Done

File: nflow-engine/src/main/java/io/nflow/engine/internal/executor/WorkflowStateProcessor.java (hunk 10 above)
Comment (efonsell) [reply nflow-engine/src/main/java/io/nflow/engine/internal/executor/WorkflowStateProcessor.java]: If started is set in db, it is never updated by instance update query

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,8 +1,12 @@
-## 5.6.1-SNAPSHOT (future release)
+## 5.7.0-SNAPSHOT (future release)
```
Comment (jsyrjala) [inline CHANGELOG.md]: Also needs a corresponding version bump to pom.xml files.

Comment (efonsell) [reply CHANGELOG.md]: Done
